README.md
# TowavePhone的博客

[![hexo version](https://img.shields.io/badge/hexo-%3E%3D%203.0-blue.svg)](http://hexo.io)
[![pipeline status](http://120.79.62.126:10080/towavephone/TowavePhoneBlog/badges/master/pipeline.svg)](http://120.79.62.126:10080/towavephone/TowavePhoneBlog/commits/master)
[![Crates.io](https://img.shields.io/crates/l/rustc-serialize.svg)]()

本博客是基于 [MOxFIVE/hexo-theme-yelee](https://github.com/MOxFIVE/hexo-theme-yelee) 制作的主题，其中添加更改了大量功能。

崇尚美观的 UI，以及极致的移动端阅读体验，你可以点击 [我的博客](http://towavephone.coding.me/) 查看效果。    

如果本主题也是你喜欢的菜，请动动手指 Star 支持一下

## 效果预览

<div align="center"><img src="./images/QQ截图20180428193728.jpg" width="400"/></div>

<div align="center"><img src="./images/QQ截图20180428193643.jpg" width="400"/></div>

## 博客改进日志

[见这里](http://towavephone.coding.me/2018/04/24/update/)     

## 使用

### Yarn

``` bash
yarn
// 因为 node_modules 文件夹里面有改动
git checkout .
yarn dev
git add .
// 输入的提交信息必须为英文
yarn cm
yarn prod
yarn deploy
```

### npm

``` bash
npm install
git checkout .
npm run dev
git add .
npm run cm
npm run prod
npm run deploy
```

## 其他

### 感谢

在设计这个博客的时候参考了好多主题和博客的设计和创意，深表感谢！

### 贡献

接受各种形式的贡献，包括但不限于提交问题或需求，修复代码，欢迎大家提 Issue 或者 Pull Request。

如果觉得本博客还不错，欢迎 Star 下，您的支持和鼓励才是后续更新最大的动力！

当然，也欢迎 fork 后创建自己的博客！



CHANGELOG.md
<a name="0.0.0"></a>
# 0.0.0 (2020-01-19)


### Features

* add acticle,change avatar,add time a few days ago ([36f29f8](https://github.com/towavephone/HexoBlog/commit/36f29f8))
* add acticle:front-end interview and optimization ([09a1216](https://github.com/towavephone/HexoBlog/commit/09a1216))
* add algorithm part one ([88ff05e](https://github.com/towavephone/HexoBlog/commit/88ff05e))
* add baidu and 360 auto push ([70becc7](https://github.com/towavephone/HexoBlog/commit/70becc7))
* add binary search tree algorithm ([e1d8c3c](https://github.com/towavephone/HexoBlog/commit/e1d8c3c))
* add computer internet part 4 ([d71d6c5](https://github.com/towavephone/HexoBlog/commit/d71d6c5))
* add computer network part 1 ([ea70cc6](https://github.com/towavephone/HexoBlog/commit/ea70cc6))
* add computer network part 2 ([969c848](https://github.com/towavephone/HexoBlog/commit/969c848))
* add computer network part 3 ([9e4f638](https://github.com/towavephone/HexoBlog/commit/9e4f638))
* add computer network part 5 ([1972556](https://github.com/towavephone/HexoBlog/commit/1972556))
* add database related ([9d4aa06](https://github.com/towavephone/HexoBlog/commit/9d4aa06))
* add deploy shell to gitlab ([002ed4e](https://github.com/towavephone/HexoBlog/commit/002ed4e))
* add design pattern part 2 ([1f7a3a5](https://github.com/towavephone/HexoBlog/commit/1f7a3a5))
* add design pattern part 3 ([4458b33](https://github.com/towavephone/HexoBlog/commit/4458b33))
* add design pattern part 4 ([556b3f3](https://github.com/towavephone/HexoBlog/commit/556b3f3))
* add design pattern part 5 ([2cc2e3c](https://github.com/towavephone/HexoBlog/commit/2cc2e3c))
* add design pattern part 6 ([f31bf0b](https://github.com/towavephone/HexoBlog/commit/f31bf0b))
* add design pattern part 7 ([fdab85e](https://github.com/towavephone/HexoBlog/commit/fdab85e))
* add design pattern part 8 ([b80230b](https://github.com/towavephone/HexoBlog/commit/b80230b))
* add gitlab shell and front end css html ([a1db2fe](https://github.com/towavephone/HexoBlog/commit/a1db2fe))
* add hash table ([c9336fc](https://github.com/towavephone/HexoBlog/commit/c9336fc))
* add hover style and code style ([b9bf526](https://github.com/towavephone/HexoBlog/commit/b9bf526))
* add html part 2 ([fe19171](https://github.com/towavephone/HexoBlog/commit/fe19171))
* add HTTP part 1 ([72f21f4](https://github.com/towavephone/HexoBlog/commit/72f21f4))
* add HTTP part 2 ([98971c8](https://github.com/towavephone/HexoBlog/commit/98971c8))
* add interview data ([51d9a3d](https://github.com/towavephone/HexoBlog/commit/51d9a3d))
* add JS and design pattern ([2bf5b16](https://github.com/towavephone/HexoBlog/commit/2bf5b16))
* add Linux shell ([1507943](https://github.com/towavephone/HexoBlog/commit/1507943))
* add live2d avatar and fix mobile slider ([2af0a59](https://github.com/towavephone/HexoBlog/commit/2af0a59))
* add markdown lint ([95ec6b9](https://github.com/towavephone/HexoBlog/commit/95ec6b9))
* add mvvm and tx interview ([64fb26f](https://github.com/towavephone/HexoBlog/commit/64fb26f))
* add Object-oriented thoughts ([c61a810](https://github.com/towavephone/HexoBlog/commit/c61a810))
* add OS principle ([c6e6b26](https://github.com/towavephone/HexoBlog/commit/c6e6b26))
* add push url to coding ([68804c3](https://github.com/towavephone/HexoBlog/commit/68804c3))
* add red black tree algorithm ([77ffd6c](https://github.com/towavephone/HexoBlog/commit/77ffd6c))
* add seek algorithm ([01bc035](https://github.com/towavephone/HexoBlog/commit/01bc035))
* add snow article ([21755f2](https://github.com/towavephone/HexoBlog/commit/21755f2))
* add sort algorithm ([d1b9fd8](https://github.com/towavephone/HexoBlog/commit/d1b9fd8))
* add You Don't Know JS part 1 ([9455f7b](https://github.com/towavephone/HexoBlog/commit/9455f7b))
* change avatar and fix copy code ([6812bc3](https://github.com/towavephone/HexoBlog/commit/6812bc3))
* change blog color style ([ae8139d](https://github.com/towavephone/HexoBlog/commit/ae8139d))
* change init wife to Neptune and add background image ([8a3eb79](https://github.com/towavephone/HexoBlog/commit/8a3eb79))
* change mobile slider ([3c628ba](https://github.com/towavephone/HexoBlog/commit/3c628ba))
* change to new site ([8feef27](https://github.com/towavephone/HexoBlog/commit/8feef27))
* delete github deploy ([e6a7840](https://github.com/towavephone/HexoBlog/commit/e6a7840))
* delete remote pic ([81b4c74](https://github.com/towavephone/HexoBlog/commit/81b4c74))
* fix backgroud image and push to github ([1def8b1](https://github.com/towavephone/HexoBlog/commit/1def8b1))
* fix encrypt style and uglify es6 ([3ee2611](https://github.com/towavephone/HexoBlog/commit/3ee2611))
* fix encrypt zip error ([844b742](https://github.com/towavephone/HexoBlog/commit/844b742))
* fix gitlab shell ([eb20468](https://github.com/towavephone/HexoBlog/commit/eb20468))
* fix gitlab shell rm error ([c284709](https://github.com/towavephone/HexoBlog/commit/c284709))
* fix hexo generate too slow and add post length ([520b3ad](https://github.com/towavephone/HexoBlog/commit/520b3ad))
* fix mobile wife load time ([06ca10b](https://github.com/towavephone/HexoBlog/commit/06ca10b))
* fix slider bug ([4d79583](https://github.com/towavephone/HexoBlog/commit/4d79583))
* fix TAG button and add front end interview html part ([e84dca0](https://github.com/towavephone/HexoBlog/commit/e84dca0))
* fix wife change bug ([337f9a8](https://github.com/towavephone/HexoBlog/commit/337f9a8))
* modify cdn url ([d0ec38a](https://github.com/towavephone/HexoBlog/commit/d0ec38a))
* modify config ([bcbd6f6](https://github.com/towavephone/HexoBlog/commit/bcbd6f6))
* modify gitalk repo ([ee3f112](https://github.com/towavephone/HexoBlog/commit/ee3f112))
* modify personal resume ([796aa98](https://github.com/towavephone/HexoBlog/commit/796aa98))
* modify personal resume part 2 and add ICP SN number ([8a4148a](https://github.com/towavephone/HexoBlog/commit/8a4148a))

gulpfile.js
var gulp = require('gulp');
var minifycss = require('gulp-minify-css');
var uglify = require('gulp-uglify-es').default;
var htmlmin = require('gulp-htmlmin');
var htmlclean = require('gulp-htmlclean');
// 压缩 public 目录 css
gulp.task('minify-css', function() {
    return gulp.src('./public/**/*.css')
        .pipe(minifycss())
        .pipe(gulp.dest('./public'));
});
// 压缩 public 目录 html
gulp.task('minify-html', function() {
  return gulp.src('./public/**/*.html')
    .pipe(htmlclean())
    .pipe(htmlmin({
         removeComments: true,
         minifyJS: true,
         minifyCSS: true,
         minifyURLs: true,
    }))
    .pipe(gulp.dest('./public'))
});
// 压缩 public/js 目录 js
gulp.task('minify-js', function() {
    return gulp.src('./public/**/*.js')
        .pipe(uglify())
        .pipe(gulp.dest('./public'));
});
// 执行 gulp 命令时执行的任务
gulp.task('default', [
    'minify-html','minify-css','minify-js'
]);


themes/yelee/source/js/pc.js
define([], function(){

    var Tips = (function(){

        var $tipBox = $(".tips-box");

        return {
            show: function(){
                $tipBox.removeClass("hide");
            },
            hide: function(){
                $tipBox.addClass("hide");
            },
            init: function(){
                
            }
        }
    })();

    var slide = function(idx){
        // 修复IE10+切换无效的bug
        var $wrap = $(".switch-wrap"),
          transform = [
              '-webkit-transform: translate(-' + idx * 100 + '%, 0);',
              '-moz-transform: translate(-' + idx * 100 + '%, 0);',
              '-o-transform: translate(-' + idx * 100 + '%, 0);',
              '-ms-transform: translate(-' + idx * 100 + '%, 0);',
              'transform: translate(-' + idx * 100 + '%, 0);'
          ];
        $wrap[0].style.cssText = transform.join('');
        $(".icon-wrap").addClass("hide");
        $(".icon-wrap").eq(idx).removeClass("hide");
    }

    var bind = function(){
        var switchBtn = $("#myonoffswitch");
        var tagcloud = $(".second-part");
        var navDiv = $(".first-part");
        switchBtn.click(function(){
            if(switchBtn.hasClass("clicked")){
                switchBtn.removeClass("clicked");
                tagcloud.removeClass("turn-left");
                navDiv.removeClass("turn-left");
            }else{
                switchBtn.addClass("clicked");
                tagcloud.addClass("turn-left");
                navDiv.addClass("turn-left");
                resetTags();
            }
        });

        var timeout;
        var isEnterBtn = false;
        var isEnterTips = false;

        $(".icon").bind("mouseenter", function(){
            isEnterBtn = true;
            Tips.show();
        }).bind("mouseleave", function(){
            isEnterBtn = false;
            setTimeout(function(){
                if(!isEnterTips){
                    Tips.hide();
                }
            }, 100);
        });

        $(".tips-box").bind("mouseenter", function(){
            isEnterTips = true;
            Tips.show();
        }).bind("mouseleave", function(){
            isEnterTips = false;
            setTimeout(function(){
                if(!isEnterBtn){
                    Tips.hide();
                }
            }, 100);
        });

        $(".tips-inner li").bind("click", function(){
            var idx = $(this).index();
            slide(idx);
            Tips.hide();
        });
    }

    var miniArchives = function(){
        if(yiliaConfig.isPost) {
            $(".post-list").addClass("toc-article");
            $("#post-nav-button > a:nth-child(2)").click(function() {
                $("#post-nav-button .fa-bars,#post-nav-button .fa-times").toggle();
                $(".post-list").toggle(300);
                if ($(".toc").length > 0) {
                    $("#toc-div, #tocButton").toggle(200, function() {
                        if ($(".switch-area").is(":visible")) {
                            $("#toc-div, .switch-btn, .switch-area").toggle();
                            $("#tocButton").attr("value", yiliaConfig.toc[0]);
                        }
                    });
                }
                else {
                    $(".switch-btn, .switch-area").fadeToggle(300);
                }
            });
        }
    }()

    if (yiliaConfig.wife) {
        (function(){
            loadCSS('/assets/waifu.css',25);
            require([yiliaConfig.rootUrl + 'assets/waifu-tips.js',yiliaConfig.rootUrl + 'assets/live2d.js'], function(){
                initModel(yiliaConfig.rootUrl + "assets/");
            })
        })()
    }

    function loadCSS (url, num) {
        var link = document.createElement("link");
        link.rel = "stylesheet";
        link.href = url;
        var head = document.querySelector("head");
        head.insertBefore(link, head.childNodes[num]);
    }

    if (yiliaConfig.jquery_ui[0]) {
        var tooltip = function(){
            require([yiliaConfig.jquery_ui[1]], function(){
                loadCSS(yiliaConfig.jquery_ui[2], 25);
                if (!$().tooltip) return;
                if (navigator.userAgent.match(/(iPhone|iPad|Android|ios|PlayBook|Touch)/i)) return;
                $("[title]").tooltip({
                    show: {
                        effect: 'blind',
                        delay: 250,
                        duration: 55,
                    }
                })
                $("#scroll").tooltip({
                    show: {
                        effect: 'clip',
                        delay: 600,
                        duration: 50,
                    }
                })
                $("#tocButton, #comments").tooltip({
                    show: {
                        delay: 1200,
                    }
                })
                $(".ds-replybox form").off("tooltip")
                $("#post-nav-button").tooltip({
                    show: {
                        effect: 'clip',
                        delay: 280,
                        duration: 65,
                    }
                })
                $("#post-nav-button > a:nth-child(2)").tooltip({
                    show: {
                        delay: 1500,
                    }
                })
                $(".social").tooltip({
                    show: {
                        effect: 'scale',
                        delay: 350,
                        duration: 70,
                    }
                })
            })
        }()
    }

    if (yiliaConfig.search) {
        var search = function(){
            require([yiliaConfig.rootUrl + 'js/search.js'], function(){
                var inputArea = document.querySelector("#local-search-input");
                var $HideWhenSearch = $("#toc-div, #tocButton, .post-list, #post-nav-button a:nth-child(2)");
                var $resetButton = $("#search-form .fa-times");
                var $resultArea = $("#local-search-result");

                var getSearchFile = function(){
                    var search_path = "search.xml";
                    var path = yiliaConfig.rootUrl + search_path;
                    searchFunc(path, 'local-search-input', 'local-search-result');
                }

                var getFileOnload = inputArea.getAttribute('searchonload');
                if (yiliaConfig.search && getFileOnload === "true") {
                    getSearchFile();
                } else {
                    inputArea.onfocus = function(){ getSearchFile() }
                }

                var HideTocArea = function(){
                    $HideWhenSearch.css("visibility","hidden");
                    $resetButton.show();
                }
                inputArea.oninput = function(){ HideTocArea() }
                inputArea.onkeydown = function(){ if(event.keyCode==13) return false}

                resetSearch = function(){
                    $HideWhenSearch.css("visibility","initial");
                    $resultArea.html("");
                    document.querySelector("#search-form").reset();
                    $resetButton.hide();
                    $(".no-result").hide();
                }

                $resultArea.bind("DOMNodeRemoved DOMNodeInserted", function(e) {
                    if (!$(e.target).text()) {
                        $(".no-result").show(200);
                    } else {
                      $(".no-result").hide();
                    }
                })
            })
        }()
    }

    return {
        init: function(){
            resetTags();
            bind();
            Tips.init();
        }
    }
});


themes/yelee/source/js/toc.js
define(function (){

    var toggleTocArea = function(){
        var valueHide = yiliaConfig.toc[0];
        var valueShow = yiliaConfig.toc[1];
        if ($(".left-col").is(":hidden")) {
            $("#tocButton").attr("value", valueHide);
        }
        $("#tocButton").click(function() {
            if ($("#toc-div").is(":hidden")) {
                $("#tocButton").attr("value", valueHide);
                $("#toc-div").slideDown(320);
                $(".switch-btn, .switch-area").fadeOut(300);
            }
            else {
                $("#tocButton").attr("value", valueShow);
                $("#toc-div").slideUp(350);
                $(".switch-btn, .switch-area").fadeIn(500);
            }
        })
    }()
    
    var HideTOCifNoHeader = function(){
        if (!$(".toc").length) {
            $("#toc-div, #tocButton").hide();
            $(".switch-btn, .switch-area").show();
        }
    }()

    $.fn.scrollUnique = function() {
        return $(this).each(function() {
            var eventType = 'mousewheel';
            // 火狐是DOMMouseScroll事件
            if (document.mozHidden !== undefined) {
                eventType = 'DOMMouseScroll';
            }
            $(this).on(eventType, function(event) {
                // 一些数据
                var scrollTop = this.scrollTop,
                    scrollHeight = this.scrollHeight,
                    height = this.clientHeight;
    
                var delta = (event.originalEvent.wheelDelta) ? event.originalEvent.wheelDelta : -(event.originalEvent.detail || 0);        
    
                if ((delta > 0 && scrollTop <= delta) || (delta < 0 && scrollHeight - height - scrollTop <= -1 * delta)) {
                    // IE浏览器下滚动会跨越边界直接影响父级滚动，因此，临界时候手动边界滚动定位
                    this.scrollTop = delta > 0? 0: scrollHeight;
                    // 向上滚 || 向下滚
                    event.preventDefault();
                }        
            });
        });	
    };

    $('#toc-div').scrollUnique();

    var $itemHasChild = $("#toc-div .toc-item:has(> .toc-child)");
    var $titleHasChild = $itemHasChild.children(".toc-link");
    $itemHasChild.prepend("<i class='fa fa-caret-down'></i><i class='fa fa-caret-right'></i>");

    var clickIcon = function(){
        $("#toc-div .toc-item > i").click(function(){
            $(this).siblings(".toc-child").slideToggle(100);
            $(this).toggleClass("hide");
            $(this).siblings("i").toggleClass("hide");
        })
    }()

    var clickTitle = function(){
        $titleHasChild.dblclick(function(){
            $(this).siblings(".toc-child").hide(100);
            $(this).siblings("i").toggleClass("hide");
        })
        // After dblclick enent
        $titleHasChild.click(function(){
            var $curentTocChild = $(this).siblings(".toc-child");
            if ($curentTocChild.is(":hidden")) {
                $curentTocChild.show(100);
                $(this).siblings("i").toggleClass("hide");
            }
        })
    }()

    var clickTocTitle = function(){
        var $iconToExpand = $(".toc-item > .fa-caret-right");
        var $iconToFold = $(".toc-item > .fa-caret-down");
        var $subToc = $titleHasChild.next(".toc-child");
        $iconToExpand.addClass("hide");

        var $tocTitle = $("#toc-div .toc-title");
        if ($titleHasChild.length) {
            $tocTitle.addClass("clickable");
            $tocTitle.click(function(){
                if ($subToc.is(":hidden")) {
                    $subToc.show(150);
                    $iconToExpand.addClass("hide");
                    $iconToFold.removeClass("hide");
                } else {
                    $subToc.hide(100);
                    $iconToExpand.removeClass("hide");
                    $iconToFold.addClass("hide");
                }
            })
            // TOC on mobile
            if ($(".left-col").is(":hidden")) {
                $("#container .toc-article .toc").css("padding-left", "1.4em");
                $("#container .toc-article .toc-title").css("display", "initial");
            }
        }
    }()

    var TocNoWarp = function(cond){
        if (cond) {
            var $tocLink = $(".toc li a");
            $tocLink.each(function(){
                var title = $(this).find('.toc-text').text();
                // Find elements with ellipsis
                if (this.offsetWidth < this.scrollWidth) {
                    $(this).attr("title", title);
                    if (!!$().tooltip) { $(this).tooltip() }
                }
            })
            var isSafari = !!navigator.userAgent.match(/Version\/[\d\.]+.*Safari/);
            if (isSafari) {
                $("#toc-div .toc-item i").css("bottom", ".1em");
            }
        }
    }
    TocNoWarp(yiliaConfig.toc[2]);

})


themes/yelee/source/js/copy.js
//此函数用于创建复制按钮
function createCopyBtns() {
    var $codeArea = $("figure table");
    //查看页面是否具有代码区域，没有代码块则不创建复制按钮
    if ($codeArea.length > 0) {
        //复制成功后将要干的事情
        function changeToSuccess(item) {
            // console.log(item);
            swal({   
                title: "",   
                text: '复制成功',  
                type: 'success',
                html: false,
                timer: 1000,   
                showConfirmButton: false
            });
        };
        //创建 全局复制按钮，仅有一组。包含：复制按钮，复制成功响应按钮
        //值得注意的是：1.按钮默认隐藏，2.位置使用绝对位置 position: absolute; (position: fixed 也可以，需要修改代码)
        $(".article-entry").before('<div id="copyBtn" style="cursor:pointer;position:absolute;top:0px;display:none;line-height: 1; font-size:1.5em"><span id="imgCopy"><i class="fa fa-paste fa-fw fa-inverse"></i></span></div>');
        //创建 复制 插件，绑定单机时间到 指定元素，支持JQuery
        var clipboard = new Clipboard('#copyBtn', {
            target: function() {
                //返回需要复制的元素内容
                return document.querySelector("[copyFlag]");
            },
            isSupported: function() {
                //支持复制内容
                return document.querySelector("[copyFlag]");
            }
        });
        //复制成功事件绑定
        clipboard.on('success',
            function(e) {
                //清除内容被选择状态
                e.clearSelection();
                changeToSuccess(e);
            });
        //复制失败绑定事件
        clipboard.on('error',
            function(e) {
                console.error('Action:', e.action);
                console.error('Trigger:', e.trigger);
            });
        $("#copyBtn").hover(
            function() {
                $("#copyBtn").css("display", "block");
            },
            function() {
                $("#copyBtn").css("display", "none");
            }
        );
    }
}
//感应鼠标是否在代码区
$("figure").hover(
    function(e) {
        //-------鼠标活动在代码块内
        //移除之前含有复制标志代码块的 copyFlag
        $("[copyFlag]").removeAttr("copyFlag");
        //在新的（当前鼠标所在代码区）代码块插入标志：copyFlag
        $(this).find(".code").attr("copyFlag", 1);
        //获取复制按钮
        $copyBtn = $("#copyBtn");
        // debugger
        if ($copyBtn.lenght != 0) {
            //获取到按钮的前提下进行一下操作
            //停止按钮动画效果
            //设置为 显示状态
            //修改 复制按钮 位置到 当前代码块开始部位
            //设置代码块 左侧位置
            $copyBtn.css("display", "block");
            $copyBtn.css("top", parseInt($copyBtn.css("top")) + $(this).offset().top - $copyBtn.offset().top + 8);
            $copyBtn.css("right", $copyBtn.width() + 5);
        }
    }, function(e){
        $("#copyBtn").css("display", "none");
    }
);

//页面载入完成后，创建复制按钮
$(document).ready(function() {
    createCopyBtns();
});

themes/yelee/source/assets/waifu-tips.js
String.prototype.render = function (context) {
    var tokenReg = /(\\)?\{([^\{\}\\]+)(\\)?\}/g;

    return this.replace(tokenReg, function (word, slash1, token, slash2) {
        if (slash1 || slash2) {  
            return word.replace('\\', '');
        }

        var variables = token.replace(/\s/g, '').split('.');
        var currentObject = context;
        var i, length, variable;

        for (i = 0, length = variables.length; i < length; ++i) {
            variable = variables[i];
            currentObject = currentObject[variable];
            if (currentObject === undefined || currentObject === null) return '';
        }
        return currentObject;
    });
};

var re = /x/;
console.log(re);
re.toString = function() {
    showMessage('哈哈，你打开了控制台，是想要看看我的秘密吗？', 5000, true);
    return '';
};

$(document).on('copy', function (){
    showMessage('你都复制了些什么呀，转载要记得加上出处哦', 5000, true);
});

$('.waifu-tool .fui-home').click(function (){
    //window.location = 'https://www.fghrsh.net/';
    window.location = window.location.protocol+'//'+window.location.hostname+'/'
});

$('.waifu-tool .fui-eye').click(function (){
    loadOtherModel();
});

$('.waifu-tool .fui-chat').click(function (){
    showHitokoto();
});

$('.waifu-tool .fui-user').click(function (){
    loadRandModel();
});

$('.waifu-tool .fui-info-circle').click(function (){
    //window.open('https://imjad.cn/archives/lab/add-dynamic-poster-girl-with-live2d-to-your-blog-02');
    window.open('https://www.fghrsh.net/post/123.html');
});

$('.waifu-tool .fui-cross').click(function (){
    showMessage('愿你有一天能与重要的人重逢', 1300, true);
    window.setTimeout(function() {
        changeWife();
    }, 1300);
});

function changeWife(){
    if(localStorage.getItem('wife') == 'live2d'){
        $('.live2d,.fui-eye,.fui-user,.fui-photo').hide();
        localStorage.setItem('wife', 'heimao');
        $('.profilepic').css('display','block');
        $('.fui-cross').attr('class','fui-heimao-cross');
    }else{
        $('.live2d,.fui-eye,.fui-user,.fui-photo').show();
        localStorage.setItem('wife', 'live2d');
        $('.profilepic').hide();
        $('.fui-heimao-cross').attr('class','fui-cross');
    }
}

function initWife(){
    if(localStorage.getItem('wife') == 'heimao'){
        $('.live2d,.fui-eye,.fui-user,.fui-photo').hide();
        $('.profilepic').css('display','block');
        $('.fui-cross').attr('class','fui-heimao-cross');
    }else{
        localStorage.setItem('wife', 'live2d');
    }
}

$('.waifu-tool .fui-photo').click(function (){
    showMessage('照好了嘛，是不是很可爱呢？', 5000, true);
    window.Live2D.captureName = 'Pio.png';
    window.Live2D.captureFrame = true;
});

(function (){
    var text;
    //var SiteIndexUrl = 'https://www.fghrsh.net/';  // 手动指定主页
    var SiteIndexUrl = window.location.protocol+'//'+window.location.hostname+'/';  // 自动获取主页
    
    if (window.location.href == SiteIndexUrl) {      // 如果是主页
        var now = (new Date()).getHours() + 1;
        if (now > 23 || now <= 5) {
            text = '你是夜猫子呀？这么晚还不睡觉，明天起的来嘛';
        } else if (now > 5 && now <= 7) {
            text = '早上好！一日之计在于晨，美好的一天就要开始了';
        } else if (now > 7 && now <= 11) {
            text = '上午好！工作顺利嘛，不要久坐，多起来走动走动哦！';
        } else if (now > 11 && now <= 13) {
            text = '中午了，工作了一个上午，现在是午餐时间！';
        } else if (now > 13 && now <= 14) {
            text = '中午了，吃了午餐了，现在是午休时间！';
        } else if (now > 14 && now <= 17) {
            text = '午后很容易犯困呢，今天的目标完成了吗？';
        } else if (now > 17 && now <= 19) {
            text = '傍晚了！窗外夕阳的景色很美丽呢，最美不过夕阳红~';
        } else if (now > 19 && now <= 21) {
            text = '晚上好，今天过得怎么样？';
        } else if (now > 21 && now <= 23) {
            text = '已经这么晚了呀，早点休息吧，晚安~';
        } else {
            text = '嗨~ 快来逗我玩吧！';
        }
    } else {
        if(document.referrer !== ''){
            var referrer = document.createElement('a');
            referrer.href = document.referrer;
            var domain = referrer.hostname.split('.')[1];
            if (window.location.hostname == referrer.hostname) {
                text = '欢迎阅读<span style="color:#0099cc;">『' + document.title.split(' - ')[0] + '』</span>';
            } else if (domain == 'baidu') {
                text = 'Hello! 来自 百度搜索 的朋友<br>你是搜索 <span style="color:#0099cc;">' + referrer.search.split('&wd=')[1].split('&')[0] + '</span> 找到的我吗？';
            } else if (domain == 'so') {
                text = 'Hello! 来自 360搜索 的朋友<br>你是搜索 <span style="color:#0099cc;">' + referrer.search.split('&q=')[1].split('&')[0] + '</span> 找到的我吗？';
            } else if (domain == 'google') {
                text = 'Hello! 来自 谷歌搜索 的朋友<br>欢迎阅读<span style="color:#0099cc;">『' + document.title.split(' - ')[0] + '』</span>';
            } else {
                text = 'Hello! 来自 <span style="color:#0099cc;">' + referrer.hostname + '</span> 的朋友';
            }
        } else {
            text = '欢迎阅读<span style="color:#0099cc;">『' + document.title.split(' - ')[0] + '』</span>';
        }
    }
    showMessage(text, 6000);
})();

// window.hitokotoTimer = window.setInterval(showHitokoto, 6000);
/* 检测用户活动状态，并在空闲时 定时显示一言 */
var getActed = false;
window.hitokotoTimer = 0;
var hitokotoInterval = false;

$(document).mousemove(function(e){getActed = true;}).keydown(function(){getActed = true;});
setInterval(function() { if (!getActed) ifActed(); else elseActed(); }, 1000);

function ifActed() {
    if (!hitokotoInterval) {
        hitokotoInterval = true;
        hitokotoTimer = window.setInterval(showHitokoto, 30000);
    }
}

function elseActed() {
    getActed = hitokotoInterval = false;
    window.clearInterval(hitokotoTimer);
}

function showHitokoto(){
    $.getJSON('//api.fghrsh.net/hitokoto/rand/?encode=jsc&uid=3335',function(result){
        var text = '这句一言出处是 <span style="color:#0099cc;">『{source}』</span>，是 <span style="color:#0099cc;">FGHRSH</span> 在 {date} 收藏的！';
        text = text.render({source: result.source, date: result.date});
        showMessage(result.hitokoto, 5000);
        window.setTimeout(function() {showMessage(text, 3000);}, 5000);
    });
}

function showMessage(text, timeout, flag){
    if(flag || sessionStorage.getItem('waifu-text') === '' || sessionStorage.getItem('waifu-text') === null){
        if(Array.isArray(text)) text = text[Math.floor(Math.random() * text.length + 1)-1];
        //console.log(text);
        
        if(flag) sessionStorage.setItem('waifu-text', text);
        
        $('.waifu-tips').stop();
        $('.waifu-tips').html(text).fadeTo(200, 1);
        if (timeout === undefined) timeout = 5000;
        hideMessage(timeout);
    }
}

function hideMessage(timeout){
    $('.waifu-tips').stop().css('opacity',1);
    if (timeout === undefined) timeout = 5000;
    window.setTimeout(function() {sessionStorage.removeItem('waifu-text')}, timeout);
    $('.waifu-tips').delay(timeout).fadeTo(200, 0);
}

function initModel(waifuPath){
    
    if (waifuPath === undefined) waifuPath = '';
    var modelId = localStorage.getItem('modelId');
    var modelTexturesId = localStorage.getItem('modelTexturesId');
    
    if (modelId == null) {
        
        /* 首次访问加载 指定模型 的 指定材质 */
        
        var modelId = 6;            // 模型 ID
        var modelTexturesId = 0    // 材质 ID
        
    } 
    loadModel(modelId, modelTexturesId);
    initWife();
	$.ajax({
        cache: true,
        url: waifuPath+'waifu-tips.json',
        dataType: "json",
        success: function (result){
            $.each(result.mouseover, function (index, tips){
                $(document).on("mouseover", tips.selector, function (){
                    var text = tips.text;
                    if(Array.isArray(tips.text)) text = tips.text[Math.floor(Math.random() * tips.text.length + 1)-1];
                    text = text.render({text: $(this).text()});
                    showMessage(text, 2000);
                });
            });
            $.each(result.click, function (index, tips){
                $(document).on("click", tips.selector, function (){
                    var text = tips.text;
                    if(Array.isArray(tips.text)) text = tips.text[Math.floor(Math.random() * tips.text.length + 1)-1];
                    text = text.render({text: $(this).text()});
                    showMessage(text, 2000, true);
                });
            });
            $.each(result.seasons, function (index, tips){
                var now = new Date();
                var after = tips.date.split('-')[0];
                var before = tips.date.split('-')[1] || after;
                
                if((after.split('/')[0] <= now.getMonth()+1 && now.getMonth()+1 <= before.split('/')[0]) && 
                   (after.split('/')[1] <= now.getDate() && now.getDate() <= before.split('/')[1])){
                    var text = tips.text;
                    if(Array.isArray(tips.text)) text = tips.text[Math.floor(Math.random() * tips.text.length + 1)-1];
                    text = text.render({year: now.getFullYear()});
                    showMessage(text, 6000, true);
                }
            });
        }
    });
}

function loadModel(modelId, modelTexturesId){
    localStorage.setItem('modelId', modelId);
    if (modelTexturesId === undefined) modelTexturesId = 0;
    localStorage.setItem('modelTexturesId', modelTexturesId);
    loadlive2d('live2d', '//api.fghrsh.net/live2d/get/?id='+modelId+'-'+modelTexturesId, console.log('live2d','模型 '+modelId+'-'+modelTexturesId+' 加载完成'));
}

function loadRandModel(){
    var modelId = localStorage.getItem('modelId');
    var modelTexturesId = localStorage.getItem('modelTexturesId');
    
    var modelTexturesRandMode = 'rand';     // 可选 'rand'(随机), 'switch'(顺序)
    
    $.ajax({
        cache: false,
        url: '//api.fghrsh.net/live2d/'+modelTexturesRandMode+'_textures/?id='+modelId+'-'+modelTexturesId,
        dataType: "json",
        success: function (result){
            if (result.textures['id'] == 1 && (modelTexturesId == 1 || modelTexturesId == 0)) {
                showMessage('我还没有其他衣服呢', 3000, true);
            } else {
                showMessage('我的新衣服好看嘛', 3000, true);
            }
            loadModel(modelId, result.textures['id']);
        }
    });
}

function loadOtherModel(){
    var modelId = localStorage.getItem('modelId');
    
    var modelTexturesRandMode = 'switch';     // 可选 'rand'(随机), 'switch'(顺序)
    
    $.ajax({
        cache: false,
        url: '//api.fghrsh.net/live2d/'+modelTexturesRandMode+'/?id='+modelId,
        dataType: "json",
        success: function (result){
            loadModel(result.model['id']);
            showMessage(result.model['message'], 3000, true);
        }
    });
}